src/events/http2.js

```diff
@@ -0,0 +1,171 @@
+            const epsagonTraceId = generateEpsagonTraceId();
+            headers['epsagon-trace-id'] = epsagonTraceId; // eslint-disable-line no-param-reassign
+
+            const resource = new serverlessEvent.Resource([
+                hostname,
+                'http',
+                headers[':method'],
+            ]);
+
+            const startTime = Date.now();
+            const httpEvent = new serverlessEvent.Event([
+                `http2-${uuid4()}`,
+                utils.createTimestampFromTime(startTime),
+                null,
+                'http',
+                0,
+                errorCode.ErrorCode.OK,
+            ]);
+
+            httpEvent.setResource(resource);
+
+            eventInterface.addToMetadata(httpEvent,
+                {
+                    http_trace_id: epsagonTraceId,
+                }, {
+                    path: headers[':path'],
+                    request_headers: reqHeaders,
+                });
+
+            clientRequest = wrappedFunction.apply(this, [headers, options]);
+            const responsePromise = new Promise((resolve) => {
+                let data = '';
+                clientRequest.on('data', (chunk) => { data += chunk; });
+
+                clientRequest.once('error', (error) => {
+                    eventInterface.setException(httpEvent, error);
+                    resolveHttpPromise(httpEvent, resolve, startTime);
+                    // if there are no listeners on eventEmitter.error, the process
+                    // should explode. let's simulate that.
+                    if (clientRequest.listenerCount('error') === 0) {
+                        throw error; // no error listener, we should explode
+                    }
+                });
+
+                clientRequest.once('close', () => {
+                    try {
+                        const responseBody = JSON.parse(data);
+                        eventInterface.addToMetadata(httpEvent, {}, {
+                            response_body: responseBody,
+                        });
+                    } catch (err) {
+                        utils.debugLog('Could not parse JSON in http2');
+                    }
+                    resolveHttpPromise(httpEvent, resolve, startTime);
+                });
+
+                clientRequest.once('response', (res) => {
+                    updateAPIGateway(res, resource, httpEvent);
+                    eventInterface.addToMetadata(httpEvent, {
+                        status: res[':status'],
+                    }, {
+                        response_headers: extractHeaders(res),
+                    });
+                });
+            }).catch((err) => {
+                tracer.addException(err);
+            });
+
+            tracer.addEvent(httpEvent, responsePromise);
+        } catch (error) {
+            tracer.addException(error);
+        }
+
+        if (!clientRequest) {
```

Comment:
If the original function throws an error, you call it twice

Reply:
👍